src/Npgsql/NpgsqlDataSourceBuilder.cs

```diff
@@ -82,7 +82,7 @@ public NpgsqlDataSourceBuilder(string? connectionString = null)
         {
             instance.AppendDefaultFactories();
             instance.AppendResolverFactory(new ExtraConversionResolverFactory());
-            instance.AppendResolverFactory(new JsonTypeInfoResolverFactory(instance.JsonSerializerOptions));
+            instance.AppendResolverFactory(() => new JsonTypeInfoResolverFactory(instance.JsonSerializerOptions));
```

Comment:
Alternatively the factory could be intrusive too
```suggestion
            instance.AppendResolverFactory(new JsonTypeInfoResolverFactory(new Lazy<JsonSerializerOptions>(() => instance.JsonSerializerOptions)));
```

And rely on the fact that the lazy nulls out the factory after materialization, so we don't root the builder forever. Though it will root Lazy<T> in NativeAOT binaries.